Demo/Controllers/MessageViewController.swift

```diff
@@ -92,22 +97,24 @@ class MessageViewController: UIViewController {
             self.present(alertController, animated: true)
         }
     }
-    
-    private func scrollToBottomIfNeeded() {
+
+    private func beforeContentChange() {
         guard let lastVisibleRow = self.tableView.indexPathsForVisibleRows?.last?.row else {
             return
         }
-        
-        let numberOfRows = self.tableView.numberOfRows(inSection: 0)
-        let lastRow = numberOfRows - 2
-        let addedRow = numberOfRows - 1
-        
-        if lastVisibleRow == lastRow || lastVisibleRow == addedRow {
-            let indexPath = IndexPath(row: addedRow, section: 0)
-            self.tableView.scrollToRow(at: indexPath, at: .bottom, animated: true)
+
+        let rowCount = self.tableView.numberOfRows(inSection: 0)
+        self.lastRowVisibleBeforeUpdate = lastVisibleRow == rowCount-1
+    }
+
+    private func afterContentChange() {
+        let rowCount = self.tableView.numberOfRows(inSection: 0)
+        let indexPath = IndexPath(row: rowCount-1, section: 0)
+        if (self.lastRowVisibleBeforeUpdate) {
+            self.tableView.scrollToRow(at: indexPath, at: .bottom, animated: false)
```

Comment:
In order to enable animations here and make the whole transition smooth we would need improve layout constraints.